_episodes/l4_04_parameterized_tests.md

```diff
@@ -0,0 +1,353 @@
+        employee.setNumberYearsEmployed(input[i]);
+        EXPECT_EQ(employee.getNetBonus(), expected_output[i]);
+    }
+}
+```
+
+On running the code with this change, we get the following output.
+
+```bash
+[==========] Running 1 test from 1 test suite.
+[----------] Global test environment set-up.
+[----------] 1 test from EmployeeTestFixture
+[ RUN      ] EmployeeTestFixture.NetBonusIsCorrectForDifferentYears
+2_Test_using_for_loop.cpp:21: Failure
+Expected equality of these values:
+  employee.getNetBonus()
+    Which is: 3000
+  expected_output[i]
+    Which is: 7000
+[  FAILED  ] EmployeeTestFixture.NetBonusIsCorrectForDifferentYears (0 ms)
+[----------] 1 test from EmployeeTestFixture (0 ms total)
+
+[----------] Global test environment tear-down
+[==========] 1 test from 1 test suite ran. (0 ms total)
+[  PASSED  ] 0 tests.
+[  FAILED  ] 1 test, listed below:
+[  FAILED  ] EmployeeTestFixture.NetBonusIsCorrectForDifferentYears
+
+ 1 FAILED TEST
+```
+
+From the output, we can clearly see that it results in complete failure of the test even though one of the conditions (or test) was right. Moreover, the output does not help much to figure out which test has exactly failed.
+
+The solution for above mentioned issues is to make use of parameterised tests and the next section describes that.
+
+## 2. Parameterised tests in GoogleTest
+
+In Google Test, parameterised tests are implemented using the `TEST_P` macro, where "P" stands for parameterised. We define a test class and then specify multiple sets of input data using the `INSTANTIATE_TEST_CASE_P` macro. Each set of input data represents a different instance of the test, and the test framework runs the test case for each instance.
+
+A parameterized tests in GoogleTest requires the following components in general.
+
+1. **A parameterised test class**: Similar to the process of a test fixture, we need to create a class derived from `testing::TestWithParam<T>` where `T` could be any valid C++ type.
+
+```cpp
+class YourTestParameterisedClass : public::TestWithParam<T> {
+    public:
+        ClassUnderTest publicInstance;
+};
+```
+
+2. **Data structure to hold your values**: We need to create some data structure to store our values (both input and output). We can use a `struct` for this purpose as shown below.
+
+```cpp
+struct MyStruct{
+    int input;
+    int output;
+    
+    //construtor of values struct
+    MyStruct(int in, int out) : input(in), output(out) {}
+};
+```
+
+Once you have defined a structure to hold your values, you can create an instance of the same with the actual set of input and output values as shown below.
+
+```cpp
+MyStruct MyValues[] = {
+    MyStruct{InputVal1, OutputVal1},  //using constructor to create an instance of MyStruct.  
+    MyStruct{InputVal2, OutputVal2}
+};
+```
+
+3. **Create you test with TEST_P macro**: Instead of `TEST_F` macro that we used for test fixture, we use a `TEST_P` macro where `P` stands for `Parameterised` as shown below.
```

Comment:
```suggestion
3. **Create your test with TEST_P macro**: Instead of `TEST_F` macro that we used for test fixture, we use a `TEST_P` macro where `P` stands for `Parameterised` as shown below.
```